CDTDatastore/touchdb/TDReplicator.m

```diff
@@ -765,13 +768,52 @@ - (void)fetchRemoteCheckpointDoc
                     if (error.code == kTDStatusNotFound) [self maybeCreateRemoteDB];
                     response = $castIf(NSDictionary, response);
                     self.remoteCheckpoint = response;
-                    NSObject* remoteLastSequence = response[@"lastSequence"];
+                    NSObject* remoteLastSequence = response[@"source_last_seq"];
+                    if (!remoteLastSequence) {
+                        // the checkpoint doc is the old format, it will be converted when we
+                        // resave, for now
+                        // fall back to the `lastSequence` property
+                        remoteLastSequence = response[@"seq"];
+                    }
 
                     if ($equal(remoteLastSequence, localLastSequence)) {
                         _lastSequence = localLastSequence;
                         CDTLogInfo(CDTREPLICATION_LOG_CONTEXT, @"%@: Replicating from lastSequence=%@",
                                 self, _lastSequence);
                     } else {
+                        // traverse the history object looking for the last session where the
+                        // session ids match.
+                        if (self.remoteCheckpoint[@"history"]) {
+                            NSArray<NSDictionary<NSString*, NSObject*>*>* remoteHistory =
+                                self.remoteCheckpoint[@"history"];
+                            NSArray<NSDictionary<NSString*, NSObject*>*>* localHistory =
+                                localCheckpoint[@"history"];
+
+                            // This assumes the  history array is ordered (most recent -> least
+                            // recent)
+                            for (NSDictionary<NSString*, NSObject*>* rHistory in remoteHistory) {
+                                NSString* sessionID = (NSString*)rHistory[@"session_id"];
+                                BOOL found = NO;
+                                for (NSDictionary<NSString*, NSObject*>* lHistory in
+                                         remoteHistory) {
+                                    if ([lHistory[@"session_id"] isEqual:sessionID]) {
+                                        found = YES;
+                                        break;
+                                    }
+                                }
+
+                                if (found) {
+                                    self.lastSequence = rHistory[@"recorded_seq"];
+                                    break;
+                                }
+                            }
+                        } else {
+                            CDTLogInfo(CDTREPLICATION_LOG_CONTEXT,
```

Comment:
@ricellis Yes I missed this one. We could put a log message in for the local checkpoint missing the history element, but since we are in control of the local doc, I think we are unlikely to encounter it missing after its been upgraded, but then again, if its unlikely to happen there is probably a customer who is going to hit it. So I'll put one in anyway.